### PR TITLE
Combine Axes.{lines,images,collections,patches,text,tables} into single list

### LINKED ISSUE #17247: Move towards making Axes.lines, Axes.patches, ... read-only views of a single child list.
Currently Axes stores child lines in Axes.lines, child patches in Axes.patches, etc.  This may be slighly inelegant compared to having a single list of children, but more importantly, #17231 revealed another problem with this design.  Currently, artists with the same zorder and the same type (line, patch, etc.) are drawn in the order in which they were inserted in their respective list (relying on the fact that `sorted` is stable, i.e. preserves original order for ties).  This is not some obscure edge case, but in fact occurs all the time when one plots multiple lines in an Axes: by default, the lines all share the same zorder and are only kept sorted form their insertion order!  However, #17231 revealed that this doesn't work in the presence of artists of two different types (in this case, a LineCollection representing errorbar "bars" and a Line2D representing the markers at the ends of the errorbars): the two artists go into separate lists (Axes.collections and Axes.lines) and then the only way to ensure that one is drawn before or after the other is to give them *different* zorders (specifically, right now the implementation of Axes.get_children always puts collections before lines in case of ties).  Different zorders sound fine, until you start having *two* errorbars, each with their bars and markers: now you want to draw in the order [bar1, marker1, bar2, marker2], but assigining different zorders to the bar and marker (but sticking to some default) will result in [bar1, bar2, marker1, marker2].

Putting all child artists in a single list would fix this problem by maintaining a single consistent insertion list.  This doesn't mean we need to get rid of Axes.lines, Axes.patches, etc.; in any case these have probably been around for too long for it to be reasonable to remove them.  However we should likely turn them into read-only lists, and then properties recomputed on-access from the children list (`tuple(art for art in self._children if isinstance(art, Line2D))`).  In fact there are other reasons for doing so; for example right now we go to great lengths to prevent reassigning artists from one Axes to another, but (I think?) one can just move an artist from one Axes' lines list to another Axes' and one will still have achieved the same result (modulo also adjusting the transform, etc.).  We simply do not want users to be able to append new artists to the list.  Another nice thing about this change is that I believe it's actually doable with a proper API deprecation period -- one "just" needs to write a proper list subclass which warns on the right methods.  I think the only use case which this change would prevent is people *deleting* artists from Axes.lines; however, I would argue that the canonical way to remove an artist is `artist.remove()`; moreover, if we really wanted to still support that we can always keep using a custom list subclass which allows that (and forwards the deletion to the master children list).

### PR DESCRIPTION
**Pull request by QuLogic**
## PR Summary

This mostly fixes #17247; combining artists into a single list has several benefits to draw ordering. You can see some of the effects in the changed test images. For simplification, the default labels for artists will now be numbered by overall children count, instead of a type-specific count.

There are a few remaining lists, but they require additional work first, or can be ignored:
* [x] `Axes.artists`; this would fix inconsistencies in `bxp`, but breaks `Colorbar` as it relies on `Axes.artists` drawing last to produce its outline.
* [x] `Axes.collections`; changes `streamplot` slightly, and `errorbar` is now correctly drawn as a group. See commit message for exact details.
* [ ] `Axes.containers`; appears to not be used in the draw tree, ~~or anything at all, really?~~ Maybe for legends only.
* [x] `Axes.images`
* [x] `Axes.lines`
* [x] `Axes.patches`
* [x] `Axes.tables`
* [x] `Axes.texts`

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [n/a] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [n/a] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/next_api_changes/* if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by dopplershift** [comment]
image
😱 

For future reviewers, that line count is only because of changes to SVG files.

**Comment by tacaswell** [comment]
I am 100% sure that this will break people who are relying on `ax.lines.clear()` to remove all of the lines.  I suspect that if we are going to go this route we are going to have to return a list sub-class that warns _a lot_ if you mutate it, but forwards those changes through as the opening pass.  We can then see much much push back we get to either move to read-only / remove it.

I'm also a bit worried about relying on type checking to re-construct the old lists, but the docs strings do name the types so I think my worries would be placated by adding a warning to those methods if the artist is of a type that would not then show up in the list.

**Comment by jklymak** [comment]
> I am 100% sure that this will break people who are relying on ax.lines.clear()

So, should these lists ever have been made public in the first place?  `ax.lines.clear()` is not going to delete the lines, it will just mean they are never drawn.  Not quite sure how we move them to being private or immutable without causing a lot of breakage.

**Comment by QuLogic** [comment]
>  `ax.lines.clear()` is not going to delete the lines,

It is pretty close though, since `Line2D._remove_method` would be `ax.lines.remove`, and the remaining stuff in `Artist.remove` is mostly about staleness:
https://github.com/matplotlib/matplotlib/blob/d9fd7f9eb11cc4369adec4a9bd01853993f8eca1/lib/matplotlib/artist.py#L163-L180

**Comment by QuLogic** [comment]
Now that #18320 is in, I can push the last step of merging `Axes.artists`. I'll work on the fancy list thing now.

**Comment by QuLogic** [comment]
Sigh, tests are fine, but doc examples trigger both `self.blah.remove` and `self.blah.append`.

**Comment by QuLogic** [comment]
I believe this is now fully working; several tests required small tweaks since things are now drawn more correctly. See the individual commit messages for further details.

File: lib/matplotlib/axes/_base.py [hunk 1]
```diff
@@ -2075,12 +2235,13 @@ def _update_patch_limits(self, patch):
...
-        Add a `~.Table` to the axes' tables; return the table.
+        Add a `~.Table` to the Axes; return the table.
         """
+        self._deprecate_noninstance('add_table', mtable.Table, tab=tab)
         self._set_artist_props(tab)
-        self.tables.append(tab)
+        self._children.append(tab)
         tab.set_clip_path(self.patch)
-        tab._remove_method = self.tables.remove
+        tab._remove_method = self._children.remove
```
Comment (dopplershift) [inline lib/matplotlib/axes/_base.py:2244]: I thought this `_remove_method` stuff was to allow automatically removing from the correct list. Is there no way we could eliminate this stuff too?

Comment (QuLogic) [reply lib/matplotlib/axes/_base.py:2244]: We could maybe add a default? But `Artist.remove` calls this, and it also is used for, e.g., removing Artists added directly to Figures, or the Legend remove, which are not linked to these lists.

**Comment by QuLogic** [comment]
> I think I actually like the old boxplot behavior of drawing the box edges over the whiskers slightly better. No way to easily preserve?

Much like the tests, we 'just' have to add the box edges to the plot later to do that. But note that this will mean that when using `patch_artist=False` (the default), the result will change (so there would be test result changes either way).

File: lib/matplotlib/axes/_base.py [hunk 2]
```diff
@@ -1180,6 +1177,148 @@ def cla(self):
...
+            )
+
+        def __repr__(self):
+            return f'<Axes.ArtistList of {self._prop_name}>'
+
+        def __len__(self):
+            return sum(self._type_check(artist)
+                       for artist in self._axes._children)
+
+        def __getitem__(self, key):
```
Comment (tacaswell) [inline lib/matplotlib/axes/_base.py:1226]: We probably should provide our own `__iter__` as well as the one we get from Sequence 

```python
    def __iter__(self):
        i = 0
        try:
            while True:
                v = self[i]
                yield v
                i += 1
        except IndexError:
            return

```

is going to have terrible line-painter issues if someone does `list(ax.lines)` with a large number of lines on the Axes.

File: lib/matplotlib/axes/_base.py [hunk 3]
```diff
@@ -859,7 +862,7 @@ def _update_transScale(self):
         self.transScale.set(
             mtransforms.blended_transform_factory(
                 self.xaxis.get_transform(), self.yaxis.get_transform()))
-        for line in getattr(self, "lines", []):  # Not set during init.
+        for line in getattr(self, "_children", []):  # Not set during init.
```
Comment (jklymak) [inline lib/matplotlib/axes/_base.py:865]: Are we Ok going through all the children instead of the lines?

File: lib/matplotlib/axes/_base.py [hunk 4]
```diff
@@ -2510,21 +2668,21 @@ def autoscale_view(self, tight=None, scalex=True, scaley=True):
...
             # Only iterate over axes and artists if needed.  The check for
-            # ``hasattr(ax, "lines")`` is necessary because this can be called
-            # very early in the axes init process (e.g., for twin axes) when
-            # these attributes don't even exist yet, in which case
+            # ``hasattr(ax, "_children")`` is necessary because this can be
+            # called very early in the axes init process (e.g., for twin axes)
+            # when these attributes don't even exist yet, in which case
             # `get_children` would raise an AttributeError.
             if self._xmargin and scalex and self._autoscaleXon:
                 x_stickies = np.sort(np.concatenate([
```
Comment (jklymak) [inline lib/matplotlib/axes/_base.py:2676]: I find this 5 lines of code completely incomprehensible.  having two for statements and an if statement in a list comprehension is too cute in my opinion.

File: lib/matplotlib/axes/_base.py (hunk 3 above)
Comment (QuLogic) [reply lib/matplotlib/axes/_base.py:865]: There is no `lines` attribute any more, only a property, which would always exist. But good point, I should probably add an `isinstance` here.

File: lib/matplotlib/axes/_base.py (hunk 1 above)
Comment (dopplershift) [reply lib/matplotlib/axes/_base.py:2251]: Ugh, I forget about artists added elsewhere. Ok, probably can't do away with `_remove_method`.

**Comment by tacaswell** [comment]
We should note that doing `ax.lines = [...]` currently _works_ (and I can imaging an defend-able use case for flipping between two sets of lines), however I do not think that this is worth the machinery to make work.  I'm inclined to document this as something that is broken (and then re-address shimming it if it causes significant user pain).

File: lib/matplotlib/axes/_base.py [hunk 5]
```diff
@@ -1225,6 +1223,153 @@ def cla(self):
...
+            return sum(self._type_check(artist)
+                       for artist in self._axes._children)
+
+        def __iter__(self):
+            for artist in self._axes._children:
+                if self._type_check(artist):
+                    yield artist
+
+        def __getitem__(self, key):
+            return [artist
```
[the remark this answers is not in the document]
Comment (QuLogic) [reply lib/matplotlib/axes/_base.py:1278]: That's why you suggested adding `__iter__`, no?

File: lib/matplotlib/axes/_axes.py [hunk 6]
```diff
@@ -868,9 +868,9 @@ def _to_points(xy1, xy2, slope):
         if line.get_clip_path() is None:
             line.set_clip_path(self.patch)
         if not line.get_label():
-            line.set_label(f"_line{len(self.lines)}")
-        self.lines.append(line)
-        line._remove_method = self.lines.remove
+            line.set_label(f"_line{len(self._children)}")
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py:871]: This changes the auto-generated label number. Do we care?

If so, we need to count the lines in children. If not we should do

```suggestion
            line.set_label(f"_child{len(self._children)}")
```

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 7]
```diff
@@ -648,10 +654,15 @@ def autoscale_view(self, tight=None, scalex=True, scaley=True,
...
-                and len(self.lines) == len(self.patches) == 0)
+            _tight = self._tight
+            if not _tight:
+                # if image data only just use the datalim
+                for artist in self._children:
+                    if isinstance(artist, mimage.AxesImage):
+                        _tight = True
+                    elif isinstance(artist, (mlines.Line2D, mpatches.Patch)):
+                        _tight = False
+                        break
```
Comment (timhoffm) [inline lib/mpl_toolkits/mplot3d/axes3d.py:665]:
```suggestion
            has_image = any(isinstance(c, mimage.AxesImage)
                            for c in self._children)
            has_line_or_patch = any(isinstance(c, (mlines.Line2D, mpatches.Patch))
                                    for c in self._children)
            _tight = self._tight or (has_image and not has_line_or_patch)
```

File: lib/matplotlib/axes/_base.py [hunk 8]
```diff
@@ -1224,6 +1222,153 @@ def cla(self):
 
         self.stale = True
 
+    """
```
Comment (timhoffm) [inline lib/matplotlib/axes/_base.py:1225]: Docstring should be within the class.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 7 above)
Comment (QuLogic) [reply lib/mpl_toolkits/mplot3d/axes3d.py:665]: Trying to avoid iterating `self._children` twice here.

File: lib/matplotlib/axes/_axes.py (hunk 6 above)
Comment (QuLogic) [reply lib/matplotlib/axes/_axes.py:871]: I don't mind changing the name, but I don't want to iterate the list again just to count how many lines there.

Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py:871]: Then, let's change the name.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 7 above)
Comment (timhoffm) [reply lib/mpl_toolkits/mplot3d/axes3d.py:665]: Why? Optimization? - I don't think that's relevant, 'cause we usually don't have thousands of Artists. And if so, iterating twice through them is probably the least of our problems. Moreover, if performance is a concern, have you profiled this? List comprehensions are better optimized in Python than equivalent for loops, so I wouldn't even know which solution is faster.

Comment (dopplershift) [reply lib/mpl_toolkits/mplot3d/axes3d.py:665]: Given the early exit condition of `any` the performance should be something measured/profiled. In the absence of that, we should go with clarity. It's not clear to me which of those that is.

File: lib/matplotlib/axes/_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tacaswell) [reply lib/matplotlib/axes/_base.py:1277]: sorry, I was being dense :sheep:

File: lib/matplotlib/axes/_axes.py (hunk 6 above)
Comment (tacaswell) [reply lib/matplotlib/axes/_axes.py:871]: If we are going to change this one to `_child` then we should change all of them (there is at least a `_image{N}`).

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 7 above)
Comment (QuLogic) [reply lib/mpl_toolkits/mplot3d/axes3d.py:657]: If I've done this correctly, I'm pretty certain that the single loop is faster than the comprehensions, whether the first condition exits early, the second does, or neither:
```python
import timeit


class AxesImage:
    pass


class Line2D:
    pass


class Patch:
    pass


class Other:
    pass


def loop(_tight, children):
    if not _tight:
        # if image data only just use the datalim
        for artist in children:
            if isinstance(artist, AxesImage):
                _tight = True
            elif isinstance(artist, (Line2D, Patch)):
                _tight = False
                break
    return _tight


def comprehension(_tight, children):
    has_image = any(isinstance(c, AxesImage) for c in children)
    has_line_or_patch = any(isinstance(c, (Line2D, Patch))
                            for c in children)
    tight = _tight or (has_image and not has_line_or_patch)
    return tight


setup = 'from __main__ import AxesImage, Line2D, Patch, Other, children'

for N in [10, 100, 1000]:
    print('N =', N)
    children = [Other()] * N
    print('No early exit')
    print('loop',
          timeit.timeit('loop(False, children)', setup=f'{setup}, loop',
                        number=1000000))
    print('comprehension',
          timeit.timeit('comprehension(False, children)',
                        setup=f'{setup}, comprehension', number=1000000))

    children = [Other()] * N
    children[1] = AxesImage()
    print('First loop early exit')
    print('loop',
          timeit.timeit('loop(False, children)', setup=f'{setup}, loop',
                        number=1000000))
    print('comprehension',
          timeit.timeit('comprehension(False, children)',
                        setup=f'{setup}, comprehension', number=1000000))

    children = [Other()] * N
    children[-2] = AxesImage()
    print('First loop late exit')
    print('loop',
          timeit.timeit('loop(False, children)', setup=f'{setup}, loop',
                        number=1000000))
    print('comprehension',
          timeit.timeit('comprehension(False, children)',
                        setup=f'{setup}, comprehension', number=1000000))

    children = [Other()] * N
    children[1] = Line2D()
    print('Second loop early exit')
    print('loop',
          timeit.timeit('loop(False, children)', setup=f'{setup}, loop',
                        number=1000000))
    print('comprehension',
          timeit.timeit('comprehension(False, children)',
                        setup=f'{setup}, comprehension', number=1000000))

    children = [Other()] * N
    children[-2] = Line2D()
    print('Second loop late exit')
    print('loop',
          timeit.timeit('loop(False, children)', setup=f'{setup}, loop',
                        number=1000000))
    print('comprehension',
          timeit.timeit('comprehension(False, children)',
                        setup=f'{setup}, comprehension', number=1000000))

    print()
    print()
```
Figure_1
I don't expect people to have 1000 artists, but the order of 10-100 _might_ be possible in 3D.

**Comment by tacaswell** [comment]
re-milestoning to 3.5 even though this is ready review.  This has a high enough chance of this causing un-foreseen consequences that we should merge it right after we branch of 3.4.x.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 9]
```diff
@@ -479,26 +482,27 @@ def do_3d_projection(artist):
...
+                collection_count = patch_count = 0
+                for artist in sorted(collections_and_patches,
+                                     key=do_3d_projection,
+                                     reverse=True):
+                    if isinstance(artist, mcoll.Collection):
+                        artist.zorder = zorder_offset + collection_count
+                        collection_count += 1
+                    elif isinstance(artist, mpatches.Patch):
+                        artist.zorder = zorder_offset + patch_count
+                        patch_count += 1
```
Comment (timhoffm) [inline lib/mpl_toolkits/mplot3d/axes3d.py:502]: Let's track zorder directly not artist counts

```suggestion
                collection_zorder = patch_zorder = zorder_offset
                for artist in sorted(collections_and_patches,
                                     key=do_3d_projection,
                                     reverse=True):
                    if isinstance(artist, mcoll.Collection):
                        artist.zorder = collection_zorder
                        collection_zorder += 1
                    elif isinstance(artist, mpatches.Patch):
                        artist.zorder = patch_zorder
                        patch_zorder += 1
```

File: lib/matplotlib/axes/_base.py [hunk 10]
```diff
@@ -1307,6 +1305,155 @@ def cla(self):
...
+
+    @property
+    def artists(self):
+        return self.ArtistList(self, 'artists', 'add_artist', invalid_types=(
+            mcoll.Collection, mimage.AxesImage, mlines.Line2D, mpatches.Patch,
+            mtable.Table, mtext.Text))
+
+    @property
+    def collections(self):
+        return self.ArtistList(self, 'collections', 'add_collection',
```
Comment (timhoffm) [inline lib/matplotlib/axes/_base.py:1429]: Wouldn't it be simpler to warn here and return a plain list compared to defining the `ArtistList` class?

Comment (QuLogic) [reply lib/matplotlib/axes/_base.py:1429]: That's how it used to be before https://github.com/matplotlib/matplotlib/pull/18216#issuecomment-673074259

Comment (timhoffm) [reply lib/matplotlib/axes/_base.py:1429]: Fair enough.

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/tests/test_transforms.py [hunk 11]
```diff
@@ -97,9 +94,8 @@ def test_pre_transform_plotting():
     u = 2*np.sin(x) + np.cos(y[:, np.newaxis])
     v = np.sin(x) - np.cos(y[:, np.newaxis])
 
-    df = 25. / 30.   # Compatibility factor for old test image
     ax.streamplot(x, y, u, v, transform=times10 + ax.transData,
-                  density=(df, df), linewidth=u**2 + v**2)
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_transforms.py:102]: Why change the density?  the diff on this image seems pretty big

**Review by tacaswell — APPROVED** [review]

File: lib/matplotlib/tests/test_transforms.py (hunk 11 above)
Comment (tacaswell) [reply lib/matplotlib/tests/test_transforms.py:102]: I was confusing old vs new and the new one looks better, so never mind on this.

**Comment by tacaswell** [comment]
anyone can merge on CI green.

**Pull request merged**